File: 202501/src/java/SlidingWindow/O3258.java
package java.SlidingWindow;

public class O3258 {
    public static void main(String[] args) {
        
    }

    public int countKConstraintSubstrings(String s, int k) {
        int left = 0, ans = 0;
        int[] cnt = new int[2];
        for (int right = 0; right < s.length(); right++) {
            int x = s.charAt(right) == '1' ? 1 : 0;
            cnt[x]++;

            while (cnt[0] > k || cnt[1] > k) {
                x = s.charAt(right) == '1' ? 1 : 0;
                cnt[x]--;
            }
            ans += right - left + 1;
        }
        return ans;
    }
}

File: 202501/src/python/SlidingWindow/O3258.py


def countKConstraintSubstrings(self, s: str, k: int) -> int:
    left = ans = 0
    cnt = [0] * 2
    for i, x in enumerate(s):
        cnt[ord(x) & 1] += 1
        while cnt[0] > k and cnt[1] > k:
            cnt[ord(s[left]) & 1] -= 1
            left += 1
        ans += i - left + 1
    return ans